// uni-utils/dist/network.js
"use strict";

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const core = require('./core');

const {
  log
} = require('./core');

const file = require('./file');

const download = async (url, option = {}) => {
  const path = option.savePath || '';
  const options = {
    timeout: 30000
  };

  if (process.env.http_proxy) {
    options.agent = new require('https-proxy-agent')(process.env.http_proxy);
  }

  if (option.hasOwnProperty('proxy')) {
    options.agent = new require('https-proxy-agent')(option.proxy);
  }

  const fetch = (...args) => Promise.resolve().then(() => _interopRequireWildcard(require('node-fetch'))).then(({
    default: fetch
  }) => fetch(...args));

  const resp = await fetch(new URL(url), options);

  if (path) {
    const pipeline = require('util').promisify(require('stream').pipeline);

    const saveFile = require('fs').createWriteStream(path);

    return pipeline(resp.body, saveFile);
  } else {
    return resp.text();
  }
};

exports.download = download;
/*
    taskList array<object{
        downUrl
        savePath
    }>
    options object{
        timeGap: 1000 下载间隔
    }
*/

const listDownload = async (taskList, options) => {
  options = {
    timeGap: 1000,
    skipExist: true,
    ...options
  };
  return core.loopTask(taskList, async task => {
    if (options.skipExist && (await file.checkFile(task.savePath))) return;

    try {
      await download(task.downUrl, {
        savePath: task.savePath
      });
    } catch (e) {
      log('文件下载失败:', e.message, task.downUrl);
      task.error = e.message;
      return task;
    }

    if (task.i) {
      log(`下载完成:${task.i}/${taskList.length}`, task.downUrl);
    }
  }, options);
};

exports.listDownload = listDownload;
